// bi/src/detect.ts
import * as fs from 'fs';
import * as path from 'path';
import { execSync } from 'child_process';

export type PackageManager = 'npm' | 'yarn' | 'pnpm' | 'bun';

export interface DetectionResult {
    pm: PackageManager;
    version: string | null;
    reason: 'lockfile' | 'packageManager-field' | 'path-lookup' | 'default';
    lockfile?: string;
}

const LOCKFILE_MAP: Array<{ file: string; pm: PackageManager }> = [
    { file: 'bun.lockb', pm: 'bun' },
    { file: 'bun.lock', pm: 'bun' },
    { file: 'pnpm-lock.yaml', pm: 'pnpm' },
    { file: 'yarn.lock', pm: 'yarn' },
    { file: 'package-lock.json', pm: 'npm' },
];

function getVersionFromField(field: string): { pm: PackageManager; version: string } | null {
    // e.g. "pnpm@8.15.0"
    const match = field.match(/^(npm|yarn|pnpm|bun)@(.+)$/);
    if (match) {
        return { pm: match[1] as PackageManager, version: match[2] };
    }
    return null;
}

function whichPM(): PackageManager | null {
    const order: PackageManager[] = ['bun', 'pnpm', 'yarn', 'npm'];
    for (const pm of order) {
        try {
            execSync(`${pm} --version`, { stdio: 'pipe', timeout: 3000 });
            return pm;
        } catch {
            // not found, try next
        }
    }
    return null;
}

function getPMVersion(pm: PackageManager): string | null {
    try {
        return execSync(`${pm} --version`, { stdio: 'pipe', timeout: 3000 }).toString().trim();
    } catch {
        return null;
    }
}

export function detectPackageManager(cwd?: string): DetectionResult {
    const dir = cwd || process.cwd();

    // 1. Check lockfiles
    for (const { file, pm } of LOCKFILE_MAP) {
        if (fs.existsSync(path.join(dir, file))) {
            return {
                pm,
                version: getPMVersion(pm),
                reason: 'lockfile',
                lockfile: file,
            };
        }
    }

    // 2. Check packageManager field in package.json
    const pkgPath = path.join(dir, 'package.json');
    if (fs.existsSync(pkgPath)) {
        try {
            const pkg = JSON.parse(fs.readFileSync(pkgPath, 'utf-8'));
            if (pkg.packageManager) {
                const parsed = getVersionFromField(pkg.packageManager);
                if (parsed) {
                    return {
                        pm: parsed.pm,
                        version: parsed.version,
                        reason: 'packageManager-field',
                    };
                }
            }
        } catch {
            // invalid package.json, continue
        }
    }

    // 3. PATH lookup
    const found = whichPM();
    if (found) {
        return {
            pm: found,
            version: getPMVersion(found),
            reason: 'path-lookup',
        };
    }

    // 4. Default to npm
    return {
        pm: 'npm',
        version: null,
        reason: 'default',
    };
}
