// bi/src/commands/exec.ts
import { detectPackageManager, PackageManager } from '../detect';
import { spawnPM, SpawnResult } from '../utils/spawn';
import { isJsonMode, printJSON, printHuman } from '../utils/output';

function buildArgs(pm: PackageManager, pkg: string, extraArgs: string[]): string[] {
    switch (pm) {
        case 'npm': return ['exec', '--', pkg, ...extraArgs];
        case 'yarn': return ['dlx', pkg, ...extraArgs];
        case 'pnpm': return ['dlx', pkg, ...extraArgs];
        case 'bun': return ['x', pkg, ...extraArgs];
    }
}

export async function exec(
    pkg: string,
    args: string[],
    options: { json?: boolean }
): Promise<{ exitCode: number; pm: string; command: string }> {
    if (!pkg) {
        printHuman('No package specified. Usage: bi x <pkg> [args...]', 'error');
        return { exitCode: 1, pm: '', command: '' };
    }

    const detection = detectPackageManager();
    const pmArgs = buildArgs(detection.pm, pkg, args);
    const command = `${detection.pm} ${pmArgs.join(' ')}`;

    printHuman(`Running: ${command}`, 'info');

    const result: SpawnResult = await spawnPM(detection.pm, pmArgs, {
        stdio: isJsonMode() ? 'pipe' : 'inherit',
    });

    const output = { exitCode: result.exitCode, pm: detection.pm, command };

    if (isJsonMode()) {
        printJSON(output);
    } else if (result.exitCode === 0) {
        printHuman('Done!', 'success');
    }

    return output;
}
