// bi/src/commands/init.ts
import { spawnPM } from '../utils/spawn';
import { spawnPM } from '../utils/spawn';
import { isJsonMode, printJSON, printHuman } from '../utils/output';

export interface Scaffold {
    name: string;
    command: string;
    description: string;
}

const SCAFFOLDS: Scaffold[] = [
    { name: 'vite', command: 'npm create vite@latest', description: 'Vite (React/Vue/Svelte/Vanilla)' },
    { name: 'next', command: 'npx create-next-app@latest', description: 'Next.js (React SSR/SSG)' },
    { name: 'vue', command: 'npm create vue@latest', description: 'Vue 3 with Vite' },
    { name: 'nuxt', command: 'npx nuxi@latest init', description: 'Nuxt 3 (Vue SSR)' },
    { name: 'svelte', command: 'npx sv create', description: 'SvelteKit' },
    { name: 'astro', command: 'npm create astro@latest', description: 'Astro (Content-focused)' },
    { name: 'express', command: 'npx express-generator', description: 'Express.js API' },
    { name: 'fastify', command: 'npm create fastify@latest', description: 'Fastify API' },
];

export async function init(
    options: { json?: boolean; template?: string }
): Promise<{ scaffold: Scaffold | null; exitCode: number }> {
    // In JSON mode, just list available scaffolds
    if (isJsonMode()) {
        if (options.template) {
            const found = SCAFFOLDS.find(s => s.name === options.template);
            if (found) {
                printJSON({ scaffold: found, exitCode: 0 });
                return { scaffold: found, exitCode: 0 };
            } else {
                printJSON({ error: `Template '${options.template}' not found`, available: SCAFFOLDS.map(s => s.name) });
                return { scaffold: null, exitCode: 1 };
            }
        }
        printJSON({ scaffolds: SCAFFOLDS });
        return { scaffold: null, exitCode: 0 };
    }

    // If template is specified, run it directly
    if (options.template) {
        const scaffold = SCAFFOLDS.find(s => s.name === options.template);
        if (!scaffold) {
            printHuman(`Template '${options.template}' not found.`, 'error');
            printHuman(`Available: ${SCAFFOLDS.map(s => s.name).join(', ')}`);
            return { scaffold: null, exitCode: 1 };
        }
        return runScaffold(scaffold);
    }

    // Interactive: list and prompt
    printHuman('\n  Available project templates:\n');
    SCAFFOLDS.forEach((s, i) => {
        printHuman(`  ${i + 1}. ${s.name.padEnd(10)} ${s.description}`);
    });
    printHuman(`\n  Use: bi init --template=<name>\n`);

    return { scaffold: null, exitCode: 0 };
}

async function runScaffold(scaffold: Scaffold): Promise<{ scaffold: Scaffold; exitCode: number }> {
    printHuman(`Scaffolding with: ${scaffold.command}`, 'info');

    const parts = scaffold.command.split(' ');
    const cmd = parts[0];
    const args = parts.slice(1);

    const result = await spawnPM(cmd, args, { stdio: 'inherit' });

    return { scaffold, exitCode: result.exitCode };
}
