// bi/src/commands/run.ts
import { detectPackageManager, PackageManager } from '../detect';
import { spawnPM, SpawnResult } from '../utils/spawn';
import { isJsonMode, printJSON, printHuman } from '../utils/output';

function buildArgs(pm: PackageManager, script: string, extraArgs: string[]): string[] {
    switch (pm) {
        case 'npm': return ['run', script, ...(extraArgs.length ? ['--', ...extraArgs] : [])];
        case 'yarn': return [script, ...extraArgs];
        case 'pnpm': return [script, ...extraArgs];
        case 'bun': return ['run', script, ...extraArgs];
    }
}

export async function run(
    script: string,
    args: string[],
    options: { json?: boolean }
): Promise<{ exitCode: number; pm: string; command: string }> {
    if (!script) {
        printHuman('No script specified. Usage: bi run <script>', 'error');
        return { exitCode: 1, pm: '', command: '' };
    }

    const detection = detectPackageManager();
    const pmArgs = buildArgs(detection.pm, script, args);
    const command = `${detection.pm} ${pmArgs.join(' ')}`;

    printHuman(`Running: ${command}`, 'info');

    const result: SpawnResult = await spawnPM(detection.pm, pmArgs, {
        stdio: isJsonMode() ? 'pipe' : 'inherit',
    });

    const output = { exitCode: result.exitCode, pm: detection.pm, command };

    if (isJsonMode()) {
        printJSON(output);
    } else if (result.exitCode === 0) {
        printHuman('Done!', 'success');
    }

    return output;
}
