// bi/src/commands/upgrade.ts
import * as fs from 'fs';
import * as path from 'path';
import * as https from 'https';
import { isJsonMode, printJSON, printHuman } from '../utils/output';

export interface OutdatedPackage {
    name: string;
    current: string;
    latest: string;
    type: 'dependencies' | 'devDependencies';
}

function fetchLatestVersion(pkg: string): Promise<string> {
    return new Promise((resolve, reject) => {
        const url = `https://registry.npmjs.org/${encodeURIComponent(pkg)}/latest`;
        https.get(url, { headers: { 'Accept': 'application/json' } }, (res) => {
            let data = '';
            res.on('data', (chunk: Buffer) => { data += chunk.toString(); });
            res.on('end', () => {
                try {
                    const json = JSON.parse(data);
                    resolve(json.version || '0.0.0');
                } catch {
                    reject(new Error(`Failed to parse response for ${pkg}`));
                }
            });
        }).on('error', reject);
    });
}

function stripRange(version: string): string {
    return version.replace(/^[\^~>=<]+/, '');
}

export async function upgrade(
    options: { json?: boolean; apply?: boolean }
): Promise<{ outdated: OutdatedPackage[]; updated: number }> {
    const pkgPath = path.join(process.cwd(), 'package.json');

    if (!fs.existsSync(pkgPath)) {
        printHuman('No package.json found in current directory.', 'error');
        return { outdated: [], updated: 0 };
    }

    const pkg = JSON.parse(fs.readFileSync(pkgPath, 'utf-8'));
    const deps = pkg.dependencies || {};
    const devDeps = pkg.devDependencies || {};

    const allDeps: Array<{ name: string; current: string; type: 'dependencies' | 'devDependencies' }> = [];

    for (const [name, version] of Object.entries(deps)) {
        allDeps.push({ name, current: version as string, type: 'dependencies' });
    }
    for (const [name, version] of Object.entries(devDeps)) {
        allDeps.push({ name, current: version as string, type: 'devDependencies' });
    }

    if (allDeps.length === 0) {
        printHuman('No dependencies found.', 'info');
        return { outdated: [], updated: 0 };
    }

    printHuman(`Checking ${allDeps.length} dependencies...`, 'info');

    const outdated: OutdatedPackage[] = [];
    const failed: string[] = [];

    // Check in batches of 5 for performance
    for (let i = 0; i < allDeps.length; i += 5) {
        const batch = allDeps.slice(i, i + 5);
        const results = await Promise.allSettled(
            batch.map(async (dep) => {
                try {
                    const latest = await fetchLatestVersion(dep.name);
                    const currentClean = stripRange(dep.current);
                    if (currentClean !== latest) {
                        outdated.push({
                            name: dep.name,
                            current: dep.current,
                            latest,
                            type: dep.type,
                        });
                    }
                } catch {
                    // Track failed packages but continue
                    failed.push(dep.name);
                }
            })
        );
    }

    // Warn about failed lookups
    if (failed.length > 0) {
        printHuman(`Warning: Could not check ${failed.length} package(s): ${failed.join(', ')}`, 'warn');
    }

    let updated = 0;

    // Apply updates if requested
    if (options.apply && outdated.length > 0) {
        for (const dep of outdated) {
            const target = dep.type === 'dependencies' ? pkg.dependencies : pkg.devDependencies;
            if (target && target[dep.name]) {
                // Preserve the original range prefix (^, ~, etc.)
                const prefix = dep.current.match(/^[\^~>=<]+/)?.[0] || '';
                target[dep.name] = `${prefix}${dep.latest}`;
                updated++;
            }
        }
        // Write updated package.json
        fs.writeFileSync(pkgPath, JSON.stringify(pkg, null, 2) + '\n');
    }

    if (isJsonMode()) {
        printJSON({ outdated, updated });
    } else if (outdated.length === 0) {
        printHuman('All dependencies are up to date!', 'success');
    } else {
        printHuman(`\n  Found ${outdated.length} outdated package(s):\n`, 'warn');
        for (const dep of outdated) {
            printHuman(`  ${dep.name}: ${dep.current} → ${dep.latest} (${dep.type})`);
        }
        printHuman('');
        if (options.apply) {
            printHuman(`Updated ${updated} package(s) in package.json.`, 'success');
        } else {
            printHuman('Run with --apply to update package.json', 'info');
        }
    }

    return { outdated, updated };
}
