// bi/src/utils/spawn.ts
import { spawn } from 'child_process';

export interface SpawnResult {
    exitCode: number;
    stdout: string;
    stderr: string;
}

export function spawnPM(
    command: string,
    args: string[],
    options?: { cwd?: string; stdio?: 'inherit' | 'pipe' }
): Promise<SpawnResult> {
    const stdio = options?.stdio || 'inherit';
    const isWindows = process.platform === 'win32';

    return new Promise((resolve, reject) => {
        const child = spawn(command, args, {
            cwd: options?.cwd || process.cwd(),
            stdio: stdio === 'pipe' ? ['pipe', 'pipe', 'pipe'] : 'inherit',
            shell: isWindows,
        });

        let stdout = '';
        let stderr = '';

        if (stdio === 'pipe') {
            child.stdout?.on('data', (data: Buffer) => { stdout += data.toString(); });
            child.stderr?.on('data', (data: Buffer) => { stderr += data.toString(); });
        }

        child.on('error', (err) => {
            reject(err);
        });

        child.on('close', (code) => {
            resolve({
                exitCode: code ?? 1,
                stdout: stdout.trim(),
                stderr: stderr.trim(),
            });
        });
    });
}
