// bi/dist/cli.js
#!/usr/bin/env node
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));

// src/cli.ts
var import_node_fs = require("fs");
var import_node_path = require("path");

// src/utils/output.ts
var _jsonMode = false;
function setJsonMode(enabled) {
  _jsonMode = enabled;
}
function isJsonMode() {
  return _jsonMode;
}
function printJSON(data) {
  process.stdout.write(JSON.stringify(data) + "\n");
}
function printHuman(message, type) {
  if (_jsonMode) return;
  const prefixes = {
    info: "\u2139",
    success: "\u2714",
    warn: "\u26A0",
    error: "\u2716"
  };
  const prefix = type ? `${prefixes[type] || ""} ` : "";
  console.log(`${prefix}${message}`);
}

// src/detect.ts
var fs = __toESM(require("fs"));
var path = __toESM(require("path"));
var import_child_process = require("child_process");
var LOCKFILE_MAP = [
  { file: "bun.lockb", pm: "bun" },
  { file: "bun.lock", pm: "bun" },
  { file: "pnpm-lock.yaml", pm: "pnpm" },
  { file: "yarn.lock", pm: "yarn" },
  { file: "package-lock.json", pm: "npm" }
];
function getVersionFromField(field) {
  const match = field.match(/^(npm|yarn|pnpm|bun)@(.+)$/);
  if (match) {
    return { pm: match[1], version: match[2] };
  }
  return null;
}
function whichPM() {
  const order = ["bun", "pnpm", "yarn", "npm"];
  for (const pm of order) {
    try {
      (0, import_child_process.execSync)(`${pm} --version`, { stdio: "pipe", timeout: 3e3 });
      return pm;
    } catch {
    }
  }
  return null;
}
function getPMVersion(pm) {
  try {
    return (0, import_child_process.execSync)(`${pm} --version`, { stdio: "pipe", timeout: 3e3 }).toString().trim();
  } catch {
    return null;
  }
}
function detectPackageManager(cwd) {
  const dir = cwd || process.cwd();
  for (const { file, pm } of LOCKFILE_MAP) {
    if (fs.existsSync(path.join(dir, file))) {
      return {
        pm,
        version: getPMVersion(pm),
        reason: "lockfile",
        lockfile: file
      };
    }
  }
  const pkgPath = path.join(dir, "package.json");
  if (fs.existsSync(pkgPath)) {
    try {
      const pkg = JSON.parse(fs.readFileSync(pkgPath, "utf-8"));
      if (pkg.packageManager) {
        const parsed = getVersionFromField(pkg.packageManager);
        if (parsed) {
          return {
            pm: parsed.pm,
            version: parsed.version,
            reason: "packageManager-field"
          };
        }
      }
    } catch {
    }
  }
  const found = whichPM();
  if (found) {
    return {
      pm: found,
      version: getPMVersion(found),
      reason: "path-lookup"
    };
  }
  return {
    pm: "npm",
    version: null,
    reason: "default"
  };
}

// src/utils/spawn.ts
var import_child_process2 = require("child_process");
function spawnPM(command, args, options) {
  const stdio = options?.stdio || "inherit";
  const isWindows = process.platform === "win32";
  return new Promise((resolve, reject) => {
    const child = (0, import_child_process2.spawn)(command, args, {
      cwd: options?.cwd || process.cwd(),
      stdio: stdio === "pipe" ? ["pipe", "pipe", "pipe"] : "inherit",
      shell: isWindows
    });
    let stdout = "";
    let stderr = "";
    if (stdio === "pipe") {
      child.stdout?.on("data", (data) => {
        stdout += data.toString();
      });
      child.stderr?.on("data", (data) => {
        stderr += data.toString();
      });
    }
    child.on("error", (err) => {
      reject(err);
    });
    child.on("close", (code) => {
      resolve({
        exitCode: code ?? 1,
        stdout: stdout.trim(),
        stderr: stderr.trim()
      });
    });
  });
}

// src/commands/install.ts
function buildArgs(pm, packages, dev, global) {
  if (packages.length === 0) {
    return ["install"];
  }
  switch (pm) {
    case "npm":
      return ["install", ...packages, ...dev ? ["-D"] : [], ...global ? ["-g"] : []];
    case "yarn":
      if (global) return ["global", "add", ...packages, ...dev ? ["-D"] : []];
      return ["add", ...packages, ...dev ? ["-D"] : []];
    case "pnpm":
      return ["add", ...packages, ...dev ? ["-D"] : [], ...global ? ["-g"] : []];
    case "bun":
      return ["add", ...packages, ...dev ? ["-d"] : [], ...global ? ["-g"] : []];
  }
}
async function install(packages, options) {
  const detection = detectPackageManager();
  const args = buildArgs(detection.pm, packages, options.dev || false, options.global || false);
  const command = `${detection.pm} ${args.join(" ")}`;
  if (!isJsonMode()) {
    printHuman(`Running: ${command}`, "info");
  }
  const result = await spawnPM(detection.pm, args, {
    stdio: isJsonMode() ? "pipe" : "inherit"
  });
  const output = { exitCode: result.exitCode, pm: detection.pm, command };
  if (isJsonMode()) {
    printJSON(output);
  } else if (result.exitCode === 0) {
    printHuman("Done!", "success");
  }
  return output;
}

// src/commands/remove.ts
function buildArgs2(pm, packages) {
  switch (pm) {
    case "npm":
      return ["uninstall", ...packages];
    case "yarn":
      return ["remove", ...packages];
    case "pnpm":
      return ["remove", ...packages];
    case "bun":
      return ["remove", ...packages];
  }
}
async function remove(packages, options) {
  if (packages.length === 0) {
    printHuman("No packages specified. Usage: bi rm <pkg...>", "error");
    return { exitCode: 1, pm: "", command: "" };
  }
  const detection = detectPackageManager();
  const args = buildArgs2(detection.pm, packages);
  const command = `${detection.pm} ${args.join(" ")}`;
  printHuman(`Running: ${command}`, "info");
  const result = await spawnPM(detection.pm, args, {
    stdio: isJsonMode() ? "pipe" : "inherit"
  });
  const output = { exitCode: result.exitCode, pm: detection.pm, command };
  if (isJsonMode()) {
    printJSON(output);
  } else if (result.exitCode === 0) {
    printHuman("Done!", "success");
  }
  return output;
}

// src/commands/run.ts
function buildArgs3(pm, script, extraArgs) {
  switch (pm) {
    case "npm":
      return ["run", script, ...extraArgs.length ? ["--", ...extraArgs] : []];
    case "yarn":
      return [script, ...extraArgs];
    case "pnpm":
      return [script, ...extraArgs];
    case "bun":
      return ["run", script, ...extraArgs];
  }
}
async function run(script, args, options) {
  if (!script) {
    printHuman("No script specified. Usage: bi run <script>", "error");
    return { exitCode: 1, pm: "", command: "" };
  }
  const detection = detectPackageManager();
  const pmArgs = buildArgs3(detection.pm, script, args);
  const command = `${detection.pm} ${pmArgs.join(" ")}`;
  printHuman(`Running: ${command}`, "info");
  const result = await spawnPM(detection.pm, pmArgs, {
    stdio: isJsonMode() ? "pipe" : "inherit"
  });
  const output = { exitCode: result.exitCode, pm: detection.pm, command };
  if (isJsonMode()) {
    printJSON(output);
  } else if (result.exitCode === 0) {
    printHuman("Done!", "success");
  }
  return output;
}

// src/commands/exec.ts
function buildArgs4(pm, pkg, extraArgs) {
  switch (pm) {
    case "npm":
      return ["exec", "--", pkg, ...extraArgs];
    case "yarn":
      return ["dlx", pkg, ...extraArgs];
    case "pnpm":
      return ["dlx", pkg, ...extraArgs];
    case "bun":
      return ["x", pkg, ...extraArgs];
  }
}
async function exec(pkg, args, options) {
  if (!pkg) {
    printHuman("No package specified. Usage: bi x <pkg> [args...]", "error");
    return { exitCode: 1, pm: "", command: "" };
  }
  const detection = detectPackageManager();
  const pmArgs = buildArgs4(detection.pm, pkg, args);
  const command = `${detection.pm} ${pmArgs.join(" ")}`;
  printHuman(`Running: ${command}`, "info");
  const result = await spawnPM(detection.pm, pmArgs, {
    stdio: isJsonMode() ? "pipe" : "inherit"
  });
  const output = { exitCode: result.exitCode, pm: detection.pm, command };
  if (isJsonMode()) {
    printJSON(output);
  } else if (result.exitCode === 0) {
    printHuman("Done!", "success");
  }
  return output;
}

// src/commands/upgrade.ts
var fs2 = __toESM(require("fs"));
var path2 = __toESM(require("path"));
var https = __toESM(require("https"));
function fetchLatestVersion(pkg) {
  return new Promise((resolve, reject) => {
    const url = `https://registry.npmjs.org/${encodeURIComponent(pkg)}/latest`;
    https.get(url, { headers: { "Accept": "application/json" } }, (res) => {
      let data = "";
      res.on("data", (chunk) => {
        data += chunk.toString();
      });
      res.on("end", () => {
        try {
          const json = JSON.parse(data);
          resolve(json.version || "0.0.0");
        } catch {
          reject(new Error(`Failed to parse response for ${pkg}`));
        }
      });
    }).on("error", reject);
  });
}
function stripRange(version) {
  return version.replace(/^[\^~>=<]+/, "");
}
async function upgrade(options) {
  const pkgPath = path2.join(process.cwd(), "package.json");
  if (!fs2.existsSync(pkgPath)) {
    printHuman("No package.json found in current directory.", "error");
    return { outdated: [], updated: 0 };
  }
  const pkg = JSON.parse(fs2.readFileSync(pkgPath, "utf-8"));
  const deps = pkg.dependencies || {};
  const devDeps = pkg.devDependencies || {};
  const allDeps = [];
  for (const [name, version] of Object.entries(deps)) {
    allDeps.push({ name, current: version, type: "dependencies" });
  }
  for (const [name, version] of Object.entries(devDeps)) {
    allDeps.push({ name, current: version, type: "devDependencies" });
  }
  if (allDeps.length === 0) {
    printHuman("No dependencies found.", "info");
    return { outdated: [], updated: 0 };
  }
  printHuman(`Checking ${allDeps.length} dependencies...`, "info");
  const outdated = [];
  const failed = [];
  for (let i = 0; i < allDeps.length; i += 5) {
    const batch = allDeps.slice(i, i + 5);
    const results = await Promise.allSettled(
      batch.map(async (dep) => {
        try {
          const latest = await fetchLatestVersion(dep.name);
          const currentClean = stripRange(dep.current);
          if (currentClean !== latest) {
            outdated.push({
              name: dep.name,
              current: dep.current,
              latest,
              type: dep.type
            });
          }
        } catch {
          failed.push(dep.name);
        }
      })
    );
  }
  if (failed.length > 0) {
    printHuman(`Warning: Could not check ${failed.length} package(s): ${failed.join(", ")}`, "warn");
  }
  let updated = 0;
  if (options.apply && outdated.length > 0) {
    for (const dep of outdated) {
      const target = dep.type === "dependencies" ? pkg.dependencies : pkg.devDependencies;
      if (target && target[dep.name]) {
        const prefix = dep.current.match(/^[\^~>=<]+/)?.[0] || "";
        target[dep.name] = `${prefix}${dep.latest}`;
        updated++;
      }
    }
    fs2.writeFileSync(pkgPath, JSON.stringify(pkg, null, 2) + "\n");
  }
  if (isJsonMode()) {
    printJSON({ outdated, updated });
  } else if (outdated.length === 0) {
    printHuman("All dependencies are up to date!", "success");
  } else {
    printHuman(`
  Found ${outdated.length} outdated package(s):
`, "warn");
    for (const dep of outdated) {
      printHuman(`  ${dep.name}: ${dep.current} \u2192 ${dep.latest} (${dep.type})`);
    }
    printHuman("");
    if (options.apply) {
      printHuman(`Updated ${updated} package(s) in package.json.`, "success");
    } else {
      printHuman("Run with --apply to update package.json", "info");
    }
  }
  return { outdated, updated };
}

// src/commands/clean.ts
var fs3 = __toESM(require("fs"));
var path3 = __toESM(require("path"));
var CLEAN_TARGETS = [
  "node_modules",
  ".cache",
  ".next",
  ".nuxt",
  "dist",
  ".turbo",
  ".parcel-cache",
  ".vite"
];
function getDirSize(dirPath) {
  let totalSize = 0;
  try {
    const entries = fs3.readdirSync(dirPath, { withFileTypes: true });
    for (const entry of entries) {
      const fullPath = path3.join(dirPath, entry.name);
      if (entry.isFile()) {
        totalSize += fs3.statSync(fullPath).size;
      } else if (entry.isDirectory()) {
        totalSize += getDirSize(fullPath);
      }
    }
  } catch {
  }
  return totalSize;
}
function formatBytes(bytes) {
  if (bytes === 0) return "0 B";
  const units = ["B", "KB", "MB", "GB"];
  const i = Math.floor(Math.log(bytes) / Math.log(1024));
  return `${(bytes / Math.pow(1024, i)).toFixed(1)} ${units[i]}`;
}
async function clean(options) {
  const cwd = options.cwd || process.cwd();
  const deleted = [];
  let freedBytes = 0;
  for (const target of CLEAN_TARGETS) {
    const fullPath = path3.join(cwd, target);
    if (fs3.existsSync(fullPath)) {
      const size = getDirSize(fullPath);
      if (options.dryRun) {
        printHuman(`  Would delete: ${target} (${formatBytes(size)})`, "info");
      } else {
        fs3.rmSync(fullPath, { recursive: true, force: true });
        printHuman(`  Deleted: ${target} (${formatBytes(size)})`, "success");
      }
      deleted.push(target);
      freedBytes += size;
    }
  }
  const result = { deleted, freedBytes };
  if (isJsonMode()) {
    printJSON(result);
  } else if (deleted.length === 0) {
    printHuman("Nothing to clean.", "info");
  } else {
    const verb = options.dryRun ? "Would free" : "Freed";
    printHuman(`
  ${verb} ${formatBytes(freedBytes)} total.`, "success");
  }
  return result;
}

// src/commands/init.ts
var SCAFFOLDS = [
  { name: "vite", command: "npm create vite@latest", description: "Vite (React/Vue/Svelte/Vanilla)" },
  { name: "next", command: "npx create-next-app@latest", description: "Next.js (React SSR/SSG)" },
  { name: "vue", command: "npm create vue@latest", description: "Vue 3 with Vite" },
  { name: "nuxt", command: "npx nuxi@latest init", description: "Nuxt 3 (Vue SSR)" },
  { name: "svelte", command: "npx sv create", description: "SvelteKit" },
  { name: "astro", command: "npm create astro@latest", description: "Astro (Content-focused)" },
  { name: "express", command: "npx express-generator", description: "Express.js API" },
  { name: "fastify", command: "npm create fastify@latest", description: "Fastify API" }
];
async function init(options) {
  if (isJsonMode()) {
    if (options.template) {
      const found = SCAFFOLDS.find((s) => s.name === options.template);
      if (found) {
        printJSON({ scaffold: found, exitCode: 0 });
        return { scaffold: found, exitCode: 0 };
      } else {
        printJSON({ error: `Template '${options.template}' not found`, available: SCAFFOLDS.map((s) => s.name) });
        return { scaffold: null, exitCode: 1 };
      }
    }
    printJSON({ scaffolds: SCAFFOLDS });
    return { scaffold: null, exitCode: 0 };
  }
  if (options.template) {
    const scaffold = SCAFFOLDS.find((s) => s.name === options.template);
    if (!scaffold) {
      printHuman(`Template '${options.template}' not found.`, "error");
      printHuman(`Available: ${SCAFFOLDS.map((s) => s.name).join(", ")}`);
      return { scaffold: null, exitCode: 1 };
    }
    return runScaffold(scaffold);
  }
  printHuman("\n  Available project templates:\n");
  SCAFFOLDS.forEach((s, i) => {
    printHuman(`  ${i + 1}. ${s.name.padEnd(10)} ${s.description}`);
  });
  printHuman(`
  Use: bi init --template=<name>
`);
  return { scaffold: null, exitCode: 0 };
}
async function runScaffold(scaffold) {
  printHuman(`Scaffolding with: ${scaffold.command}`, "info");
  const parts = scaffold.command.split(" ");
  const cmd = parts[0];
  const args = parts.slice(1);
  const result = await spawnPM(cmd, args, { stdio: "inherit" });
  return { scaffold, exitCode: result.exitCode };
}

// src/cli.ts
var VERSION = JSON.parse((0, import_node_fs.readFileSync)((0, import_node_path.join)(__dirname, "..", "package.json"), "utf8")).version;
function printHelp() {
  const detection = detectPackageManager();
  if (isJsonMode()) {
    printJSON({
      name: "bi",
      version: VERSION,
      commands: [
        { name: "i", aliases: ["install", "add"], usage: "bi i [pkg...]", desc: "Install dependencies" },
        { name: "rm", aliases: ["remove", "un", "uninstall"], usage: "bi rm <pkg...>", desc: "Remove dependencies" },
        { name: "run", aliases: ["dev", "build", "test", "start"], usage: "bi run <script>", desc: "Run a script" },
        { name: "x", aliases: ["exec"], usage: "bi x <pkg>", desc: "Execute a package" },
        { name: "up", aliases: ["upgrade", "update"], usage: "bi up", desc: "Upgrade outdated deps" },
        { name: "clean", aliases: ["purge"], usage: "bi clean", desc: "Delete node_modules and caches" },
        { name: "init", aliases: [], usage: "bi init", desc: "Scaffold a new project" }
      ],
      detected_pm: { pm: detection.pm, reason: detection.reason }
    });
    return;
  }
  console.log(`
  bi v${VERSION} \u2014 The 2-letter dev toolkit. Type less, do more.

  Usage: bi <command> [options]

  Commands:
    bi i [pkg...]         Install dependencies (aliases: install, add)
    bi rm <pkg...>        Remove dependencies (aliases: remove, un)
    bi run <script>       Run a script (shortcuts: bi dev, bi build, bi test)
    bi x <pkg> [args]     Execute a package (alias: exec)
    bi up                 Check and upgrade outdated dependencies
    bi clean              Delete node_modules and caches
    bi init               Scaffold a new project

  Global Flags:
    -h, --help            Show this help message
    -v, --version         Show version number
    --json                Output as JSON (for AI agents)

  Detected: ${detection.pm} (${detection.reason}${detection.lockfile ? `: ${detection.lockfile}` : ""})
`);
}
function parseArgs(argv) {
  const flags = {};
  const positional = [];
  for (const arg of argv) {
    if (arg === "--json") flags.json = true;
    else if (arg === "--help" || arg === "-h") flags.help = true;
    else if (arg === "--version" || arg === "-v") flags.version = true;
    else if (arg === "-D" || arg === "--dev") flags.dev = true;
    else if (arg === "-g" || arg === "--global") flags.global = true;
    else if (arg === "--dry-run") flags.dryRun = true;
    else if (arg === "--apply") flags.apply = true;
    else if (arg.startsWith("--template=")) {
      flags._template = true;
      positional.push(arg);
    } else positional.push(arg);
  }
  return {
    command: positional[0] || "",
    args: positional.slice(1),
    flags
  };
}
function extractTemplate(args) {
  for (const arg of args) {
    if (arg.startsWith("--template=")) return arg.split("=")[1];
  }
  return void 0;
}
async function main() {
  const rawArgs = process.argv.slice(2);
  const { command, args, flags } = parseArgs(rawArgs);
  if (flags.json) setJsonMode(true);
  if (flags.version) {
    if (isJsonMode()) {
      printJSON({ name: "bi", version: VERSION });
    } else {
      console.log(`bi@${VERSION}`);
    }
    process.exit(0);
  }
  if (flags.help || !command) {
    printHelp();
    process.exit(0);
  }
  const SCRIPT_SHORTCUTS = ["dev", "build", "test", "start"];
  let exitCode = 0;
  try {
    switch (command) {
      case "i":
      case "install":
      case "add": {
        const result = await install(args, { dev: flags.dev, global: flags.global });
        exitCode = result.exitCode;
        break;
      }
      case "rm":
      case "remove":
      case "un":
      case "uninstall": {
        const result = await remove(args, {});
        exitCode = result.exitCode;
        break;
      }
      case "run": {
        const result = await run(args[0] || "", args.slice(1), {});
        exitCode = result.exitCode;
        break;
      }
      case "x":
      case "exec": {
        const result = await exec(args[0] || "", args.slice(1), {});
        exitCode = result.exitCode;
        break;
      }
      case "up":
      case "upgrade":
      case "update": {
        await upgrade({ apply: flags.apply });
        break;
      }
      case "clean":
      case "purge": {
        await clean({ dryRun: flags.dryRun });
        break;
      }
      case "init": {
        const template = extractTemplate(rawArgs);
        const result = await init({ template });
        exitCode = result.exitCode;
        break;
      }
      default: {
        if (SCRIPT_SHORTCUTS.includes(command)) {
          const result = await run(command, args, {});
          exitCode = result.exitCode;
        } else {
          printHuman(`Unknown command: ${command}. Run 'bi --help'.`, "error");
          exitCode = 1;
        }
      }
    }
  } catch (err) {
    const message = err instanceof Error ? err.message : String(err);
    if (isJsonMode()) {
      printJSON({ error: message });
    } else {
      printHuman(message, "error");
    }
    exitCode = 1;
  }
  process.exit(exitCode);
}
main();
